src/providers/trusted_service/context.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use log::info;
+use mbox::MBox;
+use std::ffi::{c_void, CString};
+use std::io::{Error, ErrorKind};
+use std::ptr::null_mut;
+use ts_binding::*;
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+pub mod ts_binding {
+    include!(concat!(env!("OUT_DIR"), "/ts_bindings.rs"));
+
+    unsafe impl Send for rpc_caller {}
+    unsafe impl Sync for rpc_caller {}
+
+    unsafe impl Send for service_context {}
+    unsafe impl Sync for service_context {}
```

Comment:
@hug-dev You might've missed this when reviewing - need to do this because otherwise I can't use the provider in a multithreaded environment. We do, however, need to lock down all operations using the `Context`, otherwise Bad Things Happen ™️ .

Reply:
Is this `Context` going to be shared amoung threads? Are you going to put it in the provider?

Do the TS allow multi-threading? Could the `Context` itself implement `Send` and `Sync` so that we can put the multi-threading protections there?

Reply:
> Are you going to put it in the provider?

It already is in there: https://github.com/parallaxsecond/parsec/pull/284/files#diff-64b289133c05ec54716b4b88a821159125e8c48faf27c6df53e5a36a2f27596cR31

> Do the TS allow multi-threading?

The RPC caller isn't thread safe at the moment. We could try to get a new service context per thread/call, but I think that's not properly working yet either. So we'll have to lock somewhere in the stack.

> Could the Context itself implement Send and Sync so that we can put the multi-threading protections there?

I guess we could do that, and locking things inside the `Context` so the provider doesn't have to be aware of that.